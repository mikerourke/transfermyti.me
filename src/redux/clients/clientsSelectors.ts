import { isNil, propOr } from "ramda";

import { selectIdToLinkedId } from "~/entityOperations/selectIdToLinkedId";
import { mappingByToolNameSelector } from "~/redux/allEntities/allEntitiesSelectors";
import { createSelector, type Selector } from "~/redux/reduxTools";
import { sourceTimeEntryCountByIdFieldSelectorFactory } from "~/redux/timeEntries/timeEntriesSelectors";
import { activeWorkspaceIdSelector } from "~/redux/workspaces/workspacesSelectors";
import type {
  Client,
  ClientTableRecord,
  Mapping,
  ReduxState,
  ToolName,
} from "~/typeDefs";

export const sourceClientsByIdSelector = createSelector(
  (state: ReduxState) => state.clients.source,
  (sourceClientsById): Dictionary<Client> => sourceClientsById,
);

const targetClientsByIdSelector = createSelector(
  (state: ReduxState) => state.clients.target,
  (targetClientsById): Dictionary<Client> => targetClientsById,
);

const sourceClientsSelector = createSelector(
  sourceClientsByIdSelector,
  (sourceClients): Client[] => Object.values(sourceClients),
);

const targetClientsSelector = createSelector(
  targetClientsByIdSelector,
  (targetClientsById): Client[] => Object.values(targetClientsById),
);

export const includedSourceClientsSelector = createSelector(
  sourceClientsSelector,
  (sourceClients): Client[] =>
    sourceClients.filter((sourceClient) => sourceClient.isIncluded),
);

export const sourceClientsForTransferSelector = createSelector(
  includedSourceClientsSelector,
  (sourceClients): Client[] =>
    sourceClients.filter((sourceClient) => isNil(sourceClient.linkedId)),
);

export const sourceClientsInActiveWorkspaceSelector = createSelector(
  sourceClientsSelector,
  activeWorkspaceIdSelector,
  (sourceClients, activeWorkspaceId): Client[] =>
    sourceClients.filter(
      (sourceClient) => sourceClient.workspaceId === activeWorkspaceId,
    ),
);

export const clientIdToLinkedIdSelector = createSelector(
  sourceClientsByIdSelector,
  (sourceClientsById): Dictionary<string> =>
    selectIdToLinkedId(sourceClientsById),
);

const projectCountBySourceClientIdSelector = createSelector(
  sourceClientsSelector,
  (sourceClients): Dictionary<number> => {
    const projectCountBySourceClientId: Dictionary<number> = {};

    for (const sourceClient of sourceClients) {
      const currentCount = propOr<number, Dictionary<number>, number>(
        0,
        sourceClient.id,
        projectCountBySourceClientId,
      );

      projectCountBySourceClientId[sourceClient.id] = currentCount + 1;
    }

    return projectCountBySourceClientId;
  },
);

export const clientsForInclusionsTableSelector = createSelector(
  (state: ReduxState) => state.allEntities.areExistsInTargetShown,
  sourceClientsInActiveWorkspaceSelector,
  projectCountBySourceClientIdSelector,
  sourceTimeEntryCountByIdFieldSelectorFactory("clientId"),
  (
    areExistsInTargetShown,
    sourceClients,
    projectCountByClientId,
    timeEntryCountByClientId,
  ): ClientTableRecord[] => {
    const clientTableRecords: ClientTableRecord[] = [];

    for (const sourceClient of sourceClients) {
      const existsInTarget = sourceClient.linkedId !== null;

      if (existsInTarget && !areExistsInTargetShown) {
        continue;
      }

      const entryCount = propOr<number, Dictionary<number>, number>(
        0,
        sourceClient.id,
        timeEntryCountByClientId,
      );

      const projectCount = propOr<number, Dictionary<number>, number>(
        0,
        sourceClient.id,
        projectCountByClientId,
      );

      clientTableRecords.push({
        ...sourceClient,
        entryCount,
        projectCount,
        existsInTarget,
        isActiveInSource: true,
        isActiveInTarget: existsInTarget,
      });
    }

    return clientTableRecords;
  },
);

export const clientsTotalCountsByTypeSelector = createSelector(
  clientsForInclusionsTableSelector,
  (clientsForInclusionsTable): Dictionary<number> =>
    clientsForInclusionsTable.reduce(
      (
        acc,
        {
          entryCount,
          existsInTarget,
          projectCount,
          isIncluded,
        }: ClientTableRecord,
      ) => ({
        entryCount: acc.entryCount + entryCount,
        existsInTarget: acc.existsInTarget + (existsInTarget ? 1 : 0),
        projectCount: acc.projectCount + projectCount,
        isIncluded: acc.isIncluded + (isIncluded ? 1 : 0),
      }),
      {
        entryCount: 0,
        existsInTarget: 0,
        projectCount: 0,
        isIncluded: 0,
      },
    ),
);

const clientsByMappingSelector = createSelector(
  sourceClientsSelector,
  targetClientsSelector,
  (sourceClients, targetClients): Record<Mapping, Client[]> => ({
    source: sourceClients,
    target: targetClients,
  }),
);

export const clientIdsByNameSelectorFactory = (
  toolName: ToolName,
): Selector<ReduxState, Dictionary<string>> =>
  createSelector(
    mappingByToolNameSelector,
    clientsByMappingSelector,
    (mappingByToolName, clientsByMapping) => {
      const toolMapping = mappingByToolName[toolName];

      const clients = clientsByMapping[toolMapping];

      const clientIdsByName: Dictionary<string> = {};
      for (const client of clients) {
        clientIdsByName[client.name] = client.id;
      }

      return clientIdsByName;
    },
  );
