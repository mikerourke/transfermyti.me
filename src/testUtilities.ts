export { FAKES } from "~/redux/__fakes__/state";

type MockInstance = jest.MockInstance<AnyValid, AnyValid>;

interface WrapForMockingReturn {
  calls: AnyValid[];
  clear(): MockInstance;
  implementOnce(funcOrReturn: AnyValid): MockInstance;
  rejectAlways(err?: AnyValid): MockInstance;
  rejectOnce(err?: AnyValid): MockInstance;
  resolveMultiple(values: AnyValid[]): void;
  resolveOnce(value?: AnyValid): MockInstance;
  returnOnce(value?: AnyValid): MockInstance;
  returnMultiple(values: AnyValid[]): void;
  throwOnce(message?: string): MockInstance;
}

/**
 * Wrappers around someFunc.mock...() that Jest provides. Using
 * this avoids the annoying `someFunc as AnyValid` required to shut TypeScript
 * up.
 */
export function wrapForMocking(func: AnyValid): WrapForMockingReturn {
  return {
    calls: func?.mock?.calls,
    clear(): MockInstance {
      return func.mockClear();
    },
    implementOnce(funcOrReturn: AnyValid): MockInstance {
      if (typeof funcOrReturn === "function") {
        return func.mockImplementationOnce(funcOrReturn);
      } else {
        return func.mockImplementationOnce(() => funcOrReturn);
      }
    },
    rejectAlways(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValue(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    rejectOnce(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValueOnce(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    resolveMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockResolvedValueOnce(value);
      }
    },
    resolveOnce(value: AnyValid = null): MockInstance {
      return func.mockResolvedValueOnce(value);
    },
    returnOnce(value: AnyValid = null): MockInstance {
      return func.mockReturnValueOnce(value);
    },
    returnMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockReturnValueOnce(value);
      }
    },
    throwOnce(message: string = "error"): MockInstance {
      return func.mockImplementationOnce(() => {
        throw new Error(message);
      });
    },
  };
}

/**
 * Allows you to get around `delay()` calls in sagas.
 * @see https://github.com/jfairbank/redux-saga-test-plan/issues/257#issuecomment-503016758
 * @example
 *   return expectSaga(someSaga)
 *     .provide([
 *       { call: provideDelay },
 *     ])
 *     ...
 */
export function provideDelay(
  { fn }: AnyValid,
  next: AnyValid,
): null | VoidFunction {
  return fn.name === "delayP" ? null : next();
}
