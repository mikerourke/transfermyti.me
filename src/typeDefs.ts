import type { AllEntitiesState } from "~/redux/allEntities/allEntitiesReducer";
import type { AppState } from "~/redux/app/appReducer";
import type { ClientsState } from "~/redux/clients/clientsReducer";
import type { CredentialsState } from "~/redux/credentials/credentialsReducer";
import type { ProjectsState } from "~/redux/projects/projectsReducer";
import type { TagsState } from "~/redux/tags/tagsReducer";
import type { TasksState } from "~/redux/tasks/tasksReducer";
import type { TimeEntriesState } from "~/redux/timeEntries/timeEntriesReducer";
import type { UserGroupsState } from "~/redux/userGroups/userGroupsReducer";
import type { UsersState } from "~/redux/users/usersReducer";
import type { WorkspacesState } from "~/redux/workspaces/workspacesReducer";

export interface ReduxState {
  allEntities: AllEntitiesState;
  app: AppState;
  clients: ClientsState;
  credentials: CredentialsState;
  projects: ProjectsState;
  tags: TagsState;
  tasks: TasksState;
  timeEntries: TimeEntriesState;
  userGroups: UserGroupsState;
  users: UsersState;
  workspaces: WorkspacesState;
}

type NotificationType = "error" | "info" | "success";

export interface Notification {
  id: string;
  message: string;
  type: NotificationType;
}

export enum ToolName {
  None = "none",
  Clockify = "clockify",
  Toggl = "toggl",
}

export enum Mapping {
  Source = "source",
  Target = "target",
}

export enum ToolAction {
  None = "none",
  Delete = "delete",
  Transfer = "transfer",
}

export interface ToolHelpDetails {
  toolName: ToolName;
  displayName: string;
  toolLink: string;
}

export enum EntityGroup {
  Clients = "clients",
  Projects = "projects",
  Tags = "tags",
  Tasks = "tasks",
  TimeEntries = "timeEntries",
  UserGroups = "userGroups",
  Users = "users",
  Workspaces = "workspaces",
}

export enum FetchStatus {
  Pending = "PENDING",
  InProcess = "IN_PROCESS",
  Success = "SUCCESS",
  Error = "ERROR",
}

export interface AnyEntity {
  id: string;
  workspaceId: string;
  linkedId: string | null;
  isIncluded: boolean;
  entryCount: number;
  memberOf?: EntityGroup;
}

export type ValidEntity<TEntity> = TEntity & AnyEntity;

export type EntityTableRecord<TEntity> = TEntity & {
  existsInTarget: boolean;
  isActiveInSource: boolean;
  isActiveInTarget: boolean;
};

export type CountsByEntityGroup = Record<
  Exclude<EntityGroup, EntityGroup.Workspaces>,
  number
>;

export interface ToolNameByMapping {
  [Mapping.Source]: ToolName;
  [Mapping.Target]: ToolName;
}

export interface Client extends AnyEntity {
  id: string;
  name: string;
  workspaceId: string;
}

export type ClientTableRecord = EntityTableRecord<Client> & {
  projectCount: number;
};

export interface Credentials {
  apiKey: string | null;
  email: string | null;
  userId: string | null;
}

export type CredentialsByMapping = Record<Mapping, Credentials>;

export type ValidationErrorsByMapping = Record<Mapping, string | null>;

export interface PartialCredentialsUpdate {
  mapping: Mapping;
  apiKey?: string | null;
  email?: string | null;
  userId?: string | null;
}

export interface Estimate {
  estimate: number | string;
  type: "AUTO" | "MANUAL";
}

export interface Project extends AnyEntity {
  id: string;
  name: string;
  clientId: string | null;
  isBillable: boolean;
  isPublic: boolean;
  isActive: boolean;
  estimate: Estimate;
  color: string;
  userIds: string[];
}

export interface Tag extends AnyEntity {
  id: string;
  name: string;
  workspaceId: string;
}

export interface Task extends AnyEntity {
  id: string;
  name: string;
  estimate: string;
  projectId: string;
  assigneeIds: string[];
  isActive: boolean;
}

export type TaskTableRecord = EntityTableRecord<Task> & {
  projectName: string;
};

export interface TimeEntry extends AnyEntity {
  id: string;
  description: string;
  isBillable: boolean;
  start: Date;
  end: Date;
  year: number;
  isActive: boolean;
  clientId: string | null;
  projectId: string | null;
  tagIds: string[];
  tagNames: string[];
  taskId: string | null;
  userId: string | null;
  userGroupIds: string[];
}

export interface TimeEntryTableRecord extends EntityTableRecord<TimeEntry> {
  taskName: string | null;
  projectName: string | null;
}

export interface UserGroup extends AnyEntity {
  id: string;
  name: string;
  userIds: string[];
}

export interface User extends AnyEntity {
  id: string;
  name: string;
  email: string;
  isAdmin: boolean | null;
  isActive: boolean;
  userGroupIds: string[] | null;
}

export interface Workspace extends AnyEntity {
  id: string;
  name: string;
  userIds: string[];
  isAdmin: boolean | null;
}
