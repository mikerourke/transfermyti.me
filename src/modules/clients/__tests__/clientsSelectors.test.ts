import cases from "jest-in-case";

import { FAKES } from "~/testUtilities";
import { EntityGroup, ToolName } from "~/typeDefs";

import * as clientsSelectors from "../clientsSelectors";

const { REDUX_STATE } = FAKES;

const MOCK_STATE = {
  ...REDUX_STATE,
  clients: {
    source: {
      "3001": {
        id: "3001",
        name: "Test Client A",
        workspaceId: "1001",
        entryCount: 3,
        linkedId: "clock-client-01",
        isIncluded: false,
        memberOf: EntityGroup.Clients,
      },
      "3002": {
        id: "3002",
        name: "Test Client B",
        workspaceId: "1001",
        entryCount: 3,
        linkedId: null,
        isIncluded: false,
        memberOf: EntityGroup.Clients,
      },
      "3003": {
        id: "3003",
        name: "Test Client C",
        workspaceId: "1001",
        entryCount: 1,
        linkedId: null,
        isIncluded: true,
        memberOf: EntityGroup.Clients,
      },
    },
    target: {
      "clock-client-01": {
        id: "clock-client-01",
        name: "Test Client A",
        workspaceId: "clock-workspace-01",
        entryCount: 0,
        linkedId: "3001",
        isIncluded: false,
        memberOf: EntityGroup.Clients,
      },
    },
    isFetching: false,
  },
};

describe("within clientsSelectors", () => {
  test("the sourceClientsByIdSelector returns state.source", () => {
    const result = clientsSelectors.sourceClientsByIdSelector(MOCK_STATE);

    expect(result).toEqual(MOCK_STATE.clients.source);
  });

  cases(
    "the selectors match their snapshots",
    (options) => {
      const result = options.selector(MOCK_STATE);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the includedSourceClientsSelector",
        selector: clientsSelectors.includedSourceClientsSelector,
      },
      {
        name: "for the sourceClientsForTransferSelector",
        selector: clientsSelectors.sourceClientsForTransferSelector,
      },
      {
        name: "for the sourceClientsInActiveWorkspaceSelector",
        selector: clientsSelectors.sourceClientsInActiveWorkspaceSelector,
      },
      {
        name: "for the clientIdToLinkedIdSelector",
        selector: clientsSelectors.clientIdToLinkedIdSelector,
      },
      {
        name: "for the clientsTotalCountsByTypeSelector",
        selector: clientsSelectors.clientsTotalCountsByTypeSelector,
      },
    ],
  );

  cases(
    "the clientsForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown",
    (options) => {
      const updatedState = {
        ...MOCK_STATE,
        allEntities: {
          ...MOCK_STATE.allEntities,
          areExistsInTargetShown: options.areExistsInTargetShown,
        },
      };

      const result = clientsSelectors.clientsForInclusionsTableSelector(updatedState);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "when state.allEntities.areExistsInTargetShown = true",
        areExistsInTargetShown: true,
      },
      {
        name: "when state.allEntities.areExistsInTargetShown = false",
        areExistsInTargetShown: false,
      },
    ],
  );

  test("the clientIdsByNameSelectorFactory matches its snapshot", () => {
    const getClientIdsByName = clientsSelectors.clientIdsByNameSelectorFactory(ToolName.Toggl);

    const result = getClientIdsByName(MOCK_STATE);

    expect(result).toMatchSnapshot();
  });
});
