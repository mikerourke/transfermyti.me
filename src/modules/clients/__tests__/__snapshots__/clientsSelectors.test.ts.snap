// Vitest Snapshot v1

exports[`within clientsSelectors > the clientIdsByNameSelectorFactory matches its snapshot 1`] = `
{
  "Test Client A": "3001",
  "Test Client B": "3002",
  "Test Client C": "3003",
}
`;

exports[`within clientsSelectors > the clientsForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown > when state.allEntities.areExistsInTargetShown = false 1`] = `
[
  {
    "entryCount": 3,
    "existsInTarget": true,
    "id": "3001",
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isIncluded": false,
    "linkedId": "clock-client-01",
    "memberOf": "clients",
    "name": "Test Client A",
    "projectCount": 1,
    "workspaceId": "1001",
  },
  {
    "entryCount": 3,
    "existsInTarget": false,
    "id": "3002",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": false,
    "linkedId": null,
    "memberOf": "clients",
    "name": "Test Client B",
    "projectCount": 1,
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "3003",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "clients",
    "name": "Test Client C",
    "projectCount": 1,
    "workspaceId": "1001",
  },
]
`;

exports[`within clientsSelectors > the clientsForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown > when state.allEntities.areExistsInTargetShown = false 2`] = `
[
  {
    "entryCount": 3,
    "existsInTarget": false,
    "id": "3002",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": false,
    "linkedId": null,
    "memberOf": "clients",
    "name": "Test Client B",
    "projectCount": 1,
    "workspaceId": "1001",
  },
  {
    "entryCount": 0,
    "existsInTarget": false,
    "id": "3003",
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "clients",
    "name": "Test Client C",
    "projectCount": 1,
    "workspaceId": "1001",
  },
]
`;

exports[`within clientsSelectors > the selectors match their snapshots > for the clientsTotalCountsByTypeSelector 1`] = `
[
  {
    "entryCount": 1,
    "id": "3003",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "clients",
    "name": "Test Client C",
    "workspaceId": "1001",
  },
]
`;

exports[`within clientsSelectors > the selectors match their snapshots > for the clientsTotalCountsByTypeSelector 2`] = `
[
  {
    "entryCount": 1,
    "id": "3003",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "clients",
    "name": "Test Client C",
    "workspaceId": "1001",
  },
]
`;

exports[`within clientsSelectors > the selectors match their snapshots > for the clientsTotalCountsByTypeSelector 3`] = `
[
  {
    "entryCount": 3,
    "id": "3001",
    "isIncluded": false,
    "linkedId": "clock-client-01",
    "memberOf": "clients",
    "name": "Test Client A",
    "workspaceId": "1001",
  },
  {
    "entryCount": 3,
    "id": "3002",
    "isIncluded": false,
    "linkedId": null,
    "memberOf": "clients",
    "name": "Test Client B",
    "workspaceId": "1001",
  },
  {
    "entryCount": 1,
    "id": "3003",
    "isIncluded": true,
    "linkedId": null,
    "memberOf": "clients",
    "name": "Test Client C",
    "workspaceId": "1001",
  },
]
`;

exports[`within clientsSelectors > the selectors match their snapshots > for the clientsTotalCountsByTypeSelector 4`] = `
{
  "3001": "clock-client-01",
  "clock-client-01": "3001",
}
`;

exports[`within clientsSelectors > the selectors match their snapshots > for the clientsTotalCountsByTypeSelector 5`] = `
{
  "entryCount": 6,
  "existsInTarget": 1,
  "isIncluded": 1,
  "projectCount": 3,
}
`;
